scripts/stubsabot.py

```diff
@@ -170,6 +172,34 @@ def __str__(self) -> str:
         return f"Skipping {self.distribution}: {self.reason}"
 
 
+def all_py_files_in_source_are_in_py_typed_dirs(source: zipfile.ZipFile | tarfile.TarFile) -> bool:
+    py_typed_dirs: list[Path] = []
+    all_python_files: list[Path] = []
+    py_file_suffixes = {".py", ".pyi"}
+
+    if isinstance(source, zipfile.ZipFile):
+        path_iter = (Path(zip_info.filename) for zip_info in source.infolist() if not zip_info.is_dir())
+    else:
+        path_iter = (Path(tar_info.path) for tar_info in source if tar_info.isfile())
+
+    for path in path_iter:
+        if path.suffix in py_file_suffixes:
+            all_python_files.append(path)
+        elif path.name == "py.typed":
+            py_typed_dirs.append(path.parent)
+
+    if not py_typed_dirs:
+        return False
+    if not all_python_files:
+        return False
+
+    for path in all_python_files:
+        if not any(py_typed_dir in path.parents for py_typed_dir in py_typed_dirs):
+            return False
+    else:
+        return True
```

Comment:
nit:
```suggestion
    return True
```

Reply:
Hmm, I felt like `for`/`else` expressed the intent of the code more clearly here, but maybe that's just me :-)